### PR TITLE
Updated client to use PHP's Memcached class which includes SASL Auth support.

### PR DESCRIPTION
**Pull request by scottreese**
Since this represents a shift to using the Memcached class in lieu of `fsockopen`, I would propose this as a version 2.0 enhancement.

### CODE REVIEW

File: lib/mcstat.inc [hunk 1]
```diff
@@ -98,47 +98,40 @@ function hr_time($seconds) {
...
 {
-  private $socket_ptr = null;
+  private $mc = null;
   private $hostname = null;
   private $port = null;
+  private $username = null;
+  private $password = null;
 
-  function __construct($hostname = DEFAULT_MEMCACHE_HOST, $port = DEFAULT_MEMCACHE_PORT) {
+  function __construct($hostname, $port, $username, $password) {
```
Comment (aaronsenecal) [inline lib/mcstat.inc]: Should username and password be optional params in order to preserve backward compatibility?

File: lib/mcstat.inc [hunk 2]
```diff
@@ -150,15 +143,15 @@ class mcstat_reporter
   protected $starting_data = null;
   protected $last_data = null;
 
-  function __construct($hostname, $port) {
+  function __construct($hostname, $port, $username, $password) {
```
Comment (aaronsenecal) [inline lib/mcstat.inc]: Same here. Should we make these last two parameters optional?

File: lib/mcstat.inc (hunk 1 above)
Comment (aaronsenecal) [reply lib/mcstat.inc]: Also, is there any reason we're dropping the default values for the first two params? It looks like those values are still being used as defaults in other places in this code.

**Review by aaronsenecal — COMMENTED** [review]
Looking good! I just have a couple of concerns about backwards compatibility, but it sounds like we're planning a major version increase with this release anyway, so I'm not sure if it's really an issue.

**Review by sharz1 — APPROVED** [review]

File: lib/mcstat.inc (hunk 1 above)
Comment (scottreese) [reply lib/mcstat.inc]: The params are validated and set to appropriate defaults in the entry point for the script which is `mcstat_session::main()` starting [here](https://github.com/webbj74/mcstat/blob/ca1a6fbf7ba3949353298277ffd9d8fe92558b2d/lib/mcstat.inc#L342) so I felt that also including them in this function declaration was redundant as these values should remain immutable for the duration of the script's run.

Additionally, one of the constants in the original default doesn't actually exist. `DEFAULT_MEMCACHE_HOST` should have been `DEFAULT_MEMCACHE_HOSTNAME`. 😄

File: lib/mcstat.inc (hunk 2 above)
Comment (scottreese) [reply lib/mcstat.inc]: Same explanation as above.

**Comment by scottreese** [comment]
@aaronsenecal in this version you can still call the script with only the first two parameters (or no parameters for that matter) so things should continue to "just work" for applications which use it (i.e. the API hasn't actually changed). My thinking for making this a new major version, though, is that this introduces new functionality (i.e. being able to specify SASL Auth credentials) and also changes how the script works at a fundamental level. So the trade-off would be: use version 1.x if you want the old behavior with more stats or use version 2.x if you need to be able to specify credentials to be able to log into the memcache server.

File: lib/mcstat.inc [hunk 3]
```diff
@@ -98,47 +98,40 @@ function hr_time($seconds) {
  */
 class mcstat_client
 {
-  private $socket_ptr = null;
+  private $mc = null;
```
Comment (nhoag) [inline lib/mcstat.inc]: Preference for this property name to be spelled out, unless it's adopting a convention from other code in the repo.

**Review by nhoag — COMMENTED** [review]
> Since this represents a shift to using the Memcached class in lieu of fsockopen, I would propose this as a version 2.0 enhancement.

I may be wrong here, but I disagree that this particular change would represent a major version bump. The interface to the code has only "changed" WRT constructor signatures, which I agree with @aaronsenecal should be treated as conservatively as possible. Providing constructor param defaults would protect against the case that anyone out there is using customized callers (unlikely though it may be).

File: lib/mcstat.inc [hunk 4]
```diff
@@ -98,47 +98,40 @@ function hr_time($seconds) {
...
   private $hostname = null;
   private $port = null;
+  private $username = null;
+  private $password = null;
 
-  function __construct($hostname = DEFAULT_MEMCACHE_HOST, $port = DEFAULT_MEMCACHE_PORT) {
+  function __construct($hostname, $port, $username, $password) {
+    $this->mc = new Memcached();
```
Comment (webbj74) [inline lib/mcstat.inc]: Is it possible to add a check somewhere to ensure that the correct PHP extension is installed. IIRC there are two memcache-related PHP extensions.

**Comment by scottreese** [comment]
@nhoag I am hesitant to add defaults to every class constructor in the script because that is not how this script was set up previously. The intended use for the script is for the user to run the main script [here](https://github.com/webbj74/mcstat/blob/master/mcstat) which very explicitly calls `mcstat_session::main($argv, $argc)`. That too is the only place where any real validation of provided parameters is performed.

As for the major version bump, I felt the change was significant enough to warrant it but if we go strictly by semantic versioning, then you are right. I can go either way on it but was attempting to choose the option that would potentially cause the least pain given that the output in this version is not as extensive as in the previous version.

Version 1.1.0:
```
Memcache version 1.5.3
Max Memory Size: 128.0M  Pointer size: 64 bit

    time  cnct    used    gets    hits    miss    sets    evic   miss%   evic%      in     out
17:38:12    15  110.0M   12.6M   11.4M    1.3M  276.7k   55.4k    9.92   20.04    5.2G   48.6G

17:38:17    15  110.0M     212     188      24       6       2   11.32   33.33  134.4k    1.2M
17:38:22    15  110.0M     109      81      28       6       0   25.69    0.00  197.2k  1003.4k
17:38:27    15  110.0M    3.9k    3.8k      98      27      21    2.52   77.78  488.3k    3.4M
17:38:32    15  110.0M     100      88      12       2       1   12.00   50.00   74.8k  535.2k
^C
Session Statistics:
             Statistic             Total           Session
                uptime     3d 4h 24m 25s               20s
              libevent     2.0.21-stable                 0
           rusage_user          2m 7.64s             0.03s
         rusage_system         3m 54.51s             0.01s
       max_connections              2.0k                 0
      curr_connections                                  15
     total_connections             42.7k                 4
  rejected_connections                 0                 0
 connection_structures                30                 0
          reserved_fds                20                 0
               cmd_get             12.6M              4.3k
               cmd_set            276.7k                41
             cmd_flush                 0                 0
             cmd_touch                 0                 0
              get_hits             11.4M              4.1k
            get_misses              1.3M               162
           get_expired                 1                 0
           get_flushed                 0                 0
         delete_misses             14.9k                 2
           delete_hits            125.8k                16
           incr_misses                 0                 0
             incr_hits                 0                 0
           decr_misses                 0                 0
             decr_hits                 0                 0
            cas_misses                 0                 0
              cas_hits                 0                 0
            cas_badval                 0                 0
            touch_hits                 0                 0
          touch_misses                 0                 0
             auth_cmds                 0                 0
           auth_errors                 0                 0
            bytes_read              5.2G            894.7k
         bytes_written             48.6G              6.1M
       accepting_conns                 1                 0
   listen_disabled_num                 0                 0
time_in_listen_disabled_us                 0                 0
               threads                 4                 0
           conn_yields                 0                 0
      hash_power_level                16                 0
            hash_bytes            524.3k                 0
     hash_is_expanding                 0                 0
 slab_reassign_rescues               466                 0
slab_reassign_chunk_rescues                 0                 0
slab_reassign_evictions_nomem             17.7k                 0
slab_reassign_inline_reclaim             17.6k                 0
slab_reassign_busy_items             17.7k                 0
slab_reassign_busy_deletes                 0                 0
 slab_reassign_running                 0                 0
           slabs_moved                49                 0
   lru_crawler_running                 0                 0
    lru_crawler_starts             27.0k                 0
lru_maintainer_juggles             10.1M               751
          malloc_fails                 0                 0
    log_worker_dropped                 0                 0
    log_worker_written                 0                 0
   log_watcher_skipped                 0                 0
      log_watcher_sent                 0                 0
                 bytes            110.0M              4.7k
            curr_items                         27695-27697
           total_items            277.1k                41
 slab_global_page_pool                 0                 0
     expired_unfetched              1.1k                 0
     evicted_unfetched             44.8k                13
        evicted_active                 0                 0
             evictions             55.4k                24
             reclaimed                 7                 0
     crawler_reclaimed              1.1k                 0
 crawler_items_checked              2.8M                 0
     lrutail_reflocked              2.7k                 0
         moves_to_cold            440.9k              1.6k
         moves_to_warm            360.8k              1.6k
      moves_within_lru            665.5k              1.8k
       direct_reclaims             55.4k                24
     lru_bumps_dropped                 0                 0
Connection closed
Exiting.
```

This version:

```
Memcache version 1.5.3
Max Memory Size: 128.0M  Pointer size: 64 bit

    time  cnct    used    gets    hits    miss    sets    evic   miss%   evic%      in     out
17:40:33    15  110.0M   12.7M   11.4M    1.3M  276.9k   55.5k    9.92   20.06    5.2G   48.6G

17:40:38    15  110.0M     143     117      26       5       2   18.18   40.00  106.1k  774.8k
17:40:43    15  110.0M      30      19      11       3       0   36.67    0.00   97.1k  257.2k
17:40:48    15  110.0M     420     363      57      13       8   13.57   61.54  204.8k    2.2M
17:40:53    15  110.0M       0       0       0       0       0    0.00    0.00      24    3.2k
^C
Session Statistics:
             Statistic             Total           Session
                uptime     3d 4h 26m 46s               20s
               threads                 4                 0
   rusage_user_seconds               127                 0
rusage_user_microseconds            752.0k              4.0k
 rusage_system_seconds               234                 0
rusage_system_microseconds            652.0k             20.0k
            curr_items                         27695-27696
           total_items            277.3k                21
      curr_connections                                  15
     total_connections             42.7k                 1
 connection_structures                30                 0
                 bytes            110.0M               506
               cmd_get             12.7M               593
               cmd_set            276.9k                21
              get_hits             11.4M               499
            get_misses              1.3M                94
             evictions             55.5k                10
            bytes_read              5.2G            408.0k
         bytes_written             48.6G              3.2M
Connection closed
Exiting.
```

File: lib/mcstat.inc (hunk 4 above)
Comment (scottreese) [reply lib/mcstat.inc]: Yep, there is [Memcache](https://www.php.net/manual/en/class.memcache.php) and [Memcached](https://www.php.net/manual/en/class.memcached.php). I'll look into adding a check to make sure the extension is installed.

File: lib/mcstat.inc (hunk 3 above)
Comment (scottreese) [reply lib/mcstat.inc]: Done.

File: lib/mcstat.inc (hunk 4 above)
Comment (scottreese) [reply lib/mcstat.inc]: Added check at the top of this constructor to throw an Exception if the proper extension is not loaded.

**Comment by webbj74** [comment]
@nhoag @scottreese - I feel it definitely needs a major version bump, because the original version does not require the php memcached library to be installed on the host machine.

**Review by webbj74 — APPROVED** [review]
Thanks for adding the lib check!

**Review by pbull — APPROVED** [review]

**Review by nhoag — APPROVED** [review]

**Pull request merged**